File: cpp/3408_design_task_manager.cpp
"""
Problem: 3408 - Design Task Manager  
Description: Design a TaskManager class that supports initializing with a list of tasks (each with userId, taskId, and priority), adding new tasks, editing task priorities, removing tasks, and executing the highest-priority task. If multiple tasks share the same priority, execute the one with the smallest taskId. Return the userId of the executed task, or -1 if no valid task exists.  
Date: 2025-09-18  
"""

#include <unordered_map>
#include <queue>
#include <vector>
using namespace std;

class TaskManager {
    unordered_map<int, pair<int, int>> taskMap; // taskId → {userId, priority}
    priority_queue<tuple<int, int, int>> maxHeap; // {priority, taskId, taskId}

public:
    TaskManager(vector<vector<int>>& tasks) {
        for (auto& t : tasks) {
            add(t[0], t[1], t[2]);
        }
    }

    void add(int userId, int taskId, int priority) {
        taskMap[taskId] = {userId, priority};
        maxHeap.emplace(priority, taskId, taskId);
    }

    void edit(int taskId, int newPriority) {
        int userId = taskMap[taskId].first;
        taskMap[taskId] = {userId, newPriority};
        maxHeap.emplace(newPriority, taskId, taskId);
    }

    void rmv(int taskId) {
        taskMap.erase(taskId);
    }

    int execTop() {
        while (!maxHeap.empty()) {
            auto [priority, taskId, _] = maxHeap.top();
            maxHeap.pop();
            if (taskMap.count(taskId) && taskMap[taskId].second == priority) {
                int userId = taskMap[taskId].first;
                taskMap.erase(taskId);
                return userId;
            }
        }
        return -1;
    }
};

File: java/3408_design_task_manager.java
"""
Problem: 3408 - Design Task Manager  
Description: Design a TaskManager class that supports initializing with a list of tasks (each with userId, taskId, and priority), adding new tasks, editing task priorities, removing tasks, and executing the highest-priority task. If multiple tasks share the same priority, execute the one with the smallest taskId. Return the userId of the executed task, or -1 if no valid task exists.  
Date: 2025-09-18  
"""

import java.util.*;

class TaskManager {
    private Map<Integer, int[]> taskMap; // taskId → [userId, priority]
    private PriorityQueue<int[]> maxHeap; // [priority, taskId]

    public TaskManager(List<List<Integer>> tasks) {
        taskMap = new HashMap<>();
        maxHeap = new PriorityQueue<>((a, b) -> {
            if (b[0] != a[0]) return b[0] - a[0]; // higher priority first
            return b[1] - a[1]; // tie-breaker: higher taskId first
        });

        for (List<Integer> t : tasks) {
            add(t.get(0), t.get(1), t.get(2));
        }
    }

    public void add(int userId, int taskId, int priority) {
        taskMap.put(taskId, new int[]{userId, priority});
        maxHeap.offer(new int[]{priority, taskId});
    }

    public void edit(int taskId, int newPriority) {
        int userId = taskMap.get(taskId)[0];
        taskMap.put(taskId, new int[]{userId, newPriority});
        maxHeap.offer(new int[]{newPriority, taskId});
    }

    public void rmv(int taskId) {
        taskMap.remove(taskId);
    }

    public int execTop() {
        while (!maxHeap.isEmpty()) {
            int[] top = maxHeap.poll();
            int priority = top[0], taskId = top[1];
            if (taskMap.containsKey(taskId) && taskMap.get(taskId)[1] == priority) {
                int userId = taskMap.get(taskId)[0];
                taskMap.remove(taskId);
                return userId;
            }
        }
        return -1;
    }
}
